crates/bevy_ecs/hecs/src/query.rs

```diff
@@ -240,6 +240,85 @@ impl<'a, T: Component> Fetch<'a> for FetchMut<T> {
     }
 }
 
+macro_rules! impl_or_query {
+    ( $( $T:ident ),+; $( $t:ident ),+ ) => {
+        impl<$( $T: Query ),+> Query for Or<($( $T ),+)> {
+            type Fetch = FetchOr<($( $T::Fetch ),+)>;
+        }
+
+        impl<'a, $( $T: Fetch<'a> ),+> Fetch<'a> for FetchOr<($( $T ),+)> {
+            type Item = ($( $T::Item ),+);
+
+            fn access(archetype: &Archetype) -> Option<Access> {
+                if false $(|| $T::access(archetype).is_none() )+ {
+                    return None
+                }
+                Some(Access::Read)
+            }
+
+            fn borrow(archetype: &Archetype) {
+                $(
+                    $T::borrow(archetype);
+                 )+
+            }
+
+            unsafe fn get(archetype: &'a Archetype, offset: usize) -> Option<Self> {
+                Some(Self(( $( $T::get(archetype, offset)?),+ )))
+            }
+
+            fn release(archetype: &Archetype) {
+                $(
+                    $T::release(archetype);
+                 )+
+            }
+
+            unsafe fn next(&mut self) -> Self::Item {
+                let ($( $t ),+) = &mut self.0;
+                ($( $t.next() ),+)
+            }
+
+            unsafe fn should_skip(&self) -> bool {
+                let ($( $t ),+) = &self.0;
+                true $( && $t.should_skip() )+
+            }
+        }
+    };
+}
+
+impl_or_query!(Q1, Q2; q1, q2);
```

Comment:
Theres a trick in the macro world where you use the type identifier (ex: `Q1`) as the variable. This removes the need to pass in extra variables and counter-intuitively generally makes macros simpler / easier to read.

You just need to add this annotation to any function that uses this approach: ```[allow(non_snake_case)]```

I use this approach in bevy's into_system macros: https://github.com/bevyengine/bevy/blob/8101622f378641c90ecba5c6edaac26145a6436a/crates/bevy_ecs/src/system/into_system.rs#L157

Reply:
Thank you for the tip !